lib/pp.ml

```diff
@@ -350,6 +350,65 @@ let db_print_pp fmt pp =
 let db_string_of_pp pp =
   Format.asprintf "%a" db_print_pp pp
 
+let has_format_special s =
+  (* String.exists needs ocaml 4.13 *)
+  Seq.exists (function
+      | '@' | '%' | '\\' | '"' -> true
+      | _ -> false)
+    (String.to_seq s)
+
+let pp_as_format ?(with_tags=false) pp =
+  let open Format in
+  let fmt, return =
+    let buf = Buffer.create 200 in
+    let fmt = Format.formatter_of_buffer buf in
+    fmt, (fun () -> Format.pp_print_flush fmt (); buf)
+  in
+  let args = ref [] in
+  let open_box bty =
+    fprintf fmt "%s" "@[";
+    match bty with
+    | Pp_hbox -> fprintf fmt "<h>"
+    | Pp_vbox i -> if i = 0 then fprintf fmt "<v>" else fprintf fmt "<v %d>" i
+    | Pp_hvbox i -> if i = 0 then fprintf fmt "<hv>" else fprintf fmt "<hv %d>" i
+    | Pp_hovbox i -> if i = 0 then fprintf fmt "<hov>" else fprintf fmt "<hov %d>" i
+  in
+  let close_box () = fprintf fmt "%s" "@]" in
+  let rec pprec pp =
+  match pp with
+  | Ppcmd_empty -> ()
+  | Ppcmd_string s ->
+    if has_format_special s then begin
+      fprintf fmt "%s" "%s";
+      args := s :: !args
+    end else fprintf fmt "%s" s
```

Comment:
the current code keeps strings inline except when there's a special character (I don't guarantee that has_format_special is exhaustive currently)
we could alternatively never print strings inline, or always escape and print inline